### PR TITLE
feat: simplifies data fetching model

### PR DESCRIPTION
**Pull request by noahamac**
The data fetching model as it currently stands was the first thing I implemented and borrowed heavily from the DD model. This part of the app has disgusted me, but I actually like the solution now. 

This replaces the fetchers with a `react-query` implementation which is simple but powerful. https://react-query.tanstack.com/overview

Future work would be to eliminate the global cache entirely by refactoring QueryChart into a functional component which can use hooks. 

Still having trouble changing git branch without a page reload. Even after invalidating cache

Sorry for the big PR but many files were touched after I consolidated the Extension state 

This build of the diagram can be found here: https://demoexpo.looker.com/extensions/lookml_model_diagram::dev_ext/models/thelook_adwords/explores/events

### CODE REVIEW

File: src/utils/fetchers.ts [hunk 1]
```diff
@@ -24,179 +24,147 @@
 
  */
 
-import { useState, useEffect, useContext } from "react"
+import { useContext } from "react"
+import { useQuery, useMutation, useQueryClient } from 'react-query'
```
Comment (noahamac) [inline src/utils/fetchers.ts:28]: This file is the heart of the change; Most of the changes in this PR are resultant from consolidation of logic in this file.

File: src/utils/queries.ts [hunk 2]
```diff
@@ -1,37 +1,51 @@
...
+ import {
   ILookmlModel,
   ILookmlModelExplore,
   ILookmlModelExploreField
 } from '@looker/sdk/lib/sdk/3.1/models'
 import { exploreURL } from "./urls"
 import { Looker31SDK as LookerSDK } from '@looker/sdk/lib/sdk/3.1/methods'
-import { loadCached } from "./fetchers"
+
+const globalCache: any = {}
```
Comment (noahamac) [inline src/utils/queries.ts:29]: Moved the globalCache stuff into Queries, which are the only user of it now, as the inline queries use a class based component. Anticipating work with <Visualization /> to replace this and get onto the `react-query` model

File: src/utils/LookmlDiagrammer/diagrammer.ts [hunk 3]
```diff
@@ -174,9 +175,9 @@ export function generateExploreDiagram(explore: ILookmlModelExplore, hiddenToggl
 
   const diagrammer = new LookmlDiagrammer(diagramDict, buildOrder, explore)
   
-  const seedName = baseViewName !== "" ? baseViewName : buildOrder[0]
+  const startTableName = baseViewName !== "" ? baseViewName : buildOrder[0]
```
Comment (noahamac) [inline src/utils/LookmlDiagrammer/diagrammer.ts:178]: As promised in a previous PR

File: src/utils/LookmlDiagrammer/diagrammer.ts [hunk 4]
```diff
@@ -105,7 +105,8 @@ export function generateExploreDiagram(explore: ILookmlModelExplore, hiddenToggl
       return e.view === viewName && e.category === "dimension"
     }).length
 
-    if (explore.name === viewName || explore.sets[0].name === viewName) {
+    const firstSetName = explore.sets[0].name
+    if (explore.name === viewName || firstSetName === viewName || firstSetName.split(".")[0] === viewName) {
```
Comment (noahamac) [inline src/utils/LookmlDiagrammer/diagrammer.ts:109]: Resolves an edge case

File: src/components/Extension.tsx [hunk 5]
```diff
@@ -24,41 +24,34 @@
...
-  const [diagramError, setDiagramError] = useState<DiagramError | undefined>(undefined)
-  const [selectedBranch, setSelectedBranch] = React.useState("")
   const { modelName, exploreName, fullPage } = usePathNames()
   const [hiddenToggle, setHiddenToggle] = React.useState(VIEW_OPTIONS_HIDDEN_DEFAULT)
   const [displayFieldType, setDisplayFieldType] = React.useState(VIEW_OPTIONS_FIELDS_DEFAULT)
-  const { unfilteredModels, modelDetail, dimensions } = useSelectExplore(hiddenToggle, displayFieldType, selectedBranch, diagramError, setDiagramError)
+  const { unfilteredModels, modelDetail, dimensions } = useSelectExplore(hiddenToggle, displayFieldType)
   return (
   <ComponentsProvider themeCustomizations={{
+    fontFamilies: { brand: 'Google Sans' },
```
Comment (noahamac) [inline src/components/Extension.tsx:41]: "Google Sans" is design spec. No embedded or spartan use cases (yet!).

I suspect I won't be able to use your change until 21.8 is ESR

File: src/components/DiagramFrame/MetadataPanel/utils.ts [hunk 6]
```diff
@@ -116,11 +116,17 @@ export function getFieldMetadata(fields: ILookmlModelExploreField[], selectionIn
   }
 }
 
-export function getViewMetadata(viewResponse: any, lookmlLink: string, selectionInfo: SelectionInfoPacket) {
+export function getViewMetadata(viewResponse: ILookmlModelExplore, isLoading: boolean, lookmlLink: string, selectionInfo: SelectionInfoPacket) {
+  let sqlTableName = "unknown"
+  if (isLoading) {
+    sqlTableName = "Loading..."
```
Comment (bryans99) [inline src/components/DiagramFrame/MetadataPanel/utils.ts:122]: Overloading table name is tacky. I would prefer to seeing returning loading. Means a knock on effect in the component that uses it. From a maintainability standpoint this is bad.

File: src/components/Extension.tsx (hunk 5 above)
Comment (bryans99) [reply src/components/Extension.tsx:41]: You can, you will get an undefined value for the hostType which means you should use roboto.

File: src/utils/LookmlDiagrammer/diagrammer.ts (hunk 3 above)
Comment (bryans99) [reply src/utils/LookmlDiagrammer/diagrammer.ts:178]: Thankyou

File: src/utils/LookmlDiagrammer/utils.ts [hunk 7]
```diff
@@ -66,6 +70,7 @@
        }
      })
    })
+   views.length === 0 && views.push(exploreName)
```
Comment (bryans99) [inline src/utils/LookmlDiagrammer/utils.ts:73]: Is an if statement more readable?

File: src/utils/fetchers.ts [hunk 8]
```diff
@@ -24,179 +24,147 @@
...
   kind: string,
   message?: string
 }
 
-// user fetchers
-
-export const getMyUser = async (sdk: LookerSDK) => {
-  return sdk.ok(sdk.me())
+const defaultQueryOptions = {
+  staleTime: Infinity,
```
Comment (bryans99) [inline src/utils/fetchers.ts:46]: Infinity means keep for ever?

File: src/utils/fetchers.ts [hunk 9]
```diff
@@ -24,179 +24,147 @@
...
+  const queryClient = useQueryClient()
+  const mutation = useMutation(
+    (gitName: string) => core40SDK.ok(core40SDK.update_git_branch(projectId, {
+      name: gitName,
+      ref: ""
+    })),
+    {
+      onSuccess: () => {
+        queryClient.invalidateQueries()
+        location.reload()
```
Comment (bryans99) [inline src/utils/fetchers.ts:165]: This is not allowed in an extension (I think I mentioned this before).

File: src/utils/queries.ts (hunk 2 above)
Comment (bryans99) [reply src/utils/queries.ts:29]: Got it.

**Review by bryans99 — CHANGES_REQUESTED** [review]
Overall okay-ish considering the initial constraints. 

You've overloaded sqltable name in the metadata. Please fix that.

THis is a large PR so it is highly likely I might have missed something.

File: src/utils/LookmlDiagrammer/utils.ts (hunk 7 above)
Comment (noahamac) [reply src/utils/LookmlDiagrammer/utils.ts:73]: Yes

File: src/components/DiagramFrame/MetadataPanel/utils.ts (hunk 6 above)
Comment (noahamac) [reply src/components/DiagramFrame/MetadataPanel/utils.ts:122]: Will be addressed in follow up PR

File: src/__tests__/components/DiagramSettings.spec.tsx [hunk 10]
```diff
@@ -122,12 +122,9 @@ describe('<DiagramSettings />', () => {
...
       <DiagramSettings
       modelDetails={[{value:"modelname", label: "Model Name"}]}
-      currentModel={mockCurrentModel}
-      modelName={""}
-      setModelError={undefined}
-      setSelectedBranch={undefined}
-      branchOpts={[{value:"main", label: "Main"}]}
-      gitBranch={{name: "gitBranch"}}
+      modelDetail={{model:{},explores:[],gitBranch:{},gitBranches:[],fetchError:""}}
+      modelPathName={""}
```
[the remark this answers is not in the document]
Comment (noahamac) [reply src/__tests__/components/DiagramSettings.spec.tsx:126]: Testing the empty state. Other tests test diagram generation and display. Integration tests for interactivity. I definitely need to add more tests for the various utils more than this style of snapshot test

**Comment by noahamac** [comment]
I'm really displeased with this size of this PR. Seeking approval for merge conditional on these follow up PRs:
- reload data not page, remove `location.reload()` usage
- remove sqlTableName overloading

File: src/components/Extension.tsx [hunk 11]
```diff
@@ -24,41 +24,47 @@
 
  */
 
-import React, { useState } from "react"
-import { ComponentsProvider, MessageBar } from "@looker/components"
+import React, { useContext } from "react"
+import { ComponentsProvider } from "@looker/components"
```
Comment (bryans99) [inline src/components/Extension.tsx:28]: When we get into tree shaking, import from `@looker/components-provider`

**Review by bryans99 — APPROVED** [review]
This is okay for now

**Pull request merged**